### PR TITLE
maxAge: Fix logic with number + use seconds instead of ms

### PR DESCRIPTION
**Pull request by ziluvatar**
Fixes [#346](https://github.com/auth0/node-jsonwebtoken/issues/346)
Fixes [#273](https://github.com/auth0/node-jsonwebtoken/issues/273).

Since this will expect seconds to be used (even using string type), I mark it for `next-major`.

### CODE REVIEW

File: verify.js [hunk 1]
```diff
@@ -165,15 +165,13 @@ module.exports = function (jwtString, secretOrPublicKey, options, callback) {
...
     if (typeof payload.iat !== 'number') {
       return done(new JsonWebTokenError('iat required when maxAge is specified'));
     }
-    // We have to compare against either options.clockTimestamp or the currentDate _with_ millis
-    // to not change behaviour (version 7.2.1). Should be resolve somehow for next major.
-    var nowOrClockTimestamp = ((options.clockTimestamp || 0) * 1000) || Date.now();
-    if (nowOrClockTimestamp - (payload.iat * 1000) > maxAge + (options.clockTolerance || 0) * 1000) {
-      return done(new TokenExpiredError('maxAge exceeded', new Date(payload.iat * 1000 + maxAge)));
+
+    var maxAgeTimestamp = timespan(options.maxAge, payload.iat);
```
Comment (mborst) [inline verify.js]: Are you aware that your if `options.maxAge` was created by a `new String(...)` call, this will always return an error? Your `timespan` method does a `typeof x === 'string'` but `typeof new String('foo') === object`.

Comment (ziluvatar) [inline verify.js]: I didn't think about that case, I reused one function we already have. I'll take a look to it.

Comment (mborst) [inline verify.js]: I just stumbled over it when looking at the implementation of that function. I guess different rules apply now since it has to deal with external input.

Comment (ziluvatar) [inline verify.js]: I just took the same approach used in `sign()` https://github.com/auth0/node-jsonwebtoken/blob/master/sign.js#L114-L116 so that we can cover when the format of the string itself is not correct.

Comment (mborst) [inline verify.js]: I think it would be nicer for the user to generally test for string with something like
https://github.com/lodash/lodash/blob/master/isString.js, but that might better be done in a different PR.

Comment (ziluvatar) [inline verify.js]: In this case `ms(new String('1s'))` gives you `Error: val is not a non-empty string or a valid number. val="1s"`. For other cases where it does not matter one or the other it makes totally sense. Since we are entering lodash type checks in other PR, we could check&modify the ones for strings with `isString`.

Comment (mborst) [inline verify.js]: Ah, ok, so your way of doing it is 👍

Comment (dinvlad) [inline verify.js]: Has it been/planned to be merged?

Comment (dinvlad) [inline verify.js]: Also, why not simply ms(ms(options.maxAge))?

Comment (ziluvatar) [inline verify.js]: I wanted to merge a bunch of "breaking changes" and then increase major version. It seems creators of the others PRs forgot about them (we asked for changes).
So, I'll try to find time to fix them probably, but I don't have ETA for now.

> why not simply ms(ms(options.maxAge))?

ms(ms(number)) => number
ms(ms(string)) => string
We want end up having the number.

Comment (dinvlad) [reply verify.js]: Ok thanks, in that case ms(ms(ms())) might also do the trick...

Comment (dinvlad) [reply verify.js]: Ok, thanks, got it.

Comment (dinvlad) [reply verify.js]: To add to the conversation above, both `ms(String('1s'))` and `ms(String(1000))` work. No need to use new, as [String](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String) object can be used as a global constructor for strings.

Comment (ziluvatar) [reply verify.js]: Well, it is the consumer who is deciding what to send. But thanks to your comment I realized I didn't add a test for `String('12s')`, I'll add it later.

Comment (dinvlad) [reply verify.js]: Makes sense - thanks for noting that!

File: test/verify.tests.js [hunk 2]
```diff
@@ -114,70 +114,97 @@ describe('verify', function() {
...
+            assert.isUndefined(p);
+            done();
+          });
         });
       });
 
-      it('should not error for claims issued before a certain timespan but still inside clockTolerance timespan', function (done) {
-        clock = sinon.useFakeTimers(1437018582500);
-        var options = {algorithms: ['HS256'], maxAge: '321ms', clockTolerance: 100};
+      [String('6s'), '5s', 5].forEach(function (maxAge) {
```
Comment (fiddur) [inline test/verify.tests.js]: All the other multi-tests have different representations of the same time; this is different.  Seems like a mistake?

Comment (ziluvatar) [reply test/verify.tests.js]: Yep, copy&paste mistake. Fixing.
